File: src-tauri/src/discovery/subproject_registry.rs
/// Subproject registry — tracks composite project IDs for multi-cwd projects.
///
/// When sessions in the same encoded directory have different `cwd` values,
/// they are split into separate projects with composite IDs: `{encodedPath}::{hash}`.

use std::collections::{HashMap, HashSet};

use sha2::{Digest, Sha256};

/// Entry in the subproject registry.
#[derive(Debug, Clone)]
pub struct SubprojectEntry {
    pub cwd: String,
    pub session_ids: HashSet<String>,
}

/// Registry of composite project IDs.
#[derive(Debug, Default)]
pub struct SubprojectRegistry {
    entries: HashMap<String, SubprojectEntry>,
}

impl SubprojectRegistry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Register a subproject and return its composite ID.
    pub fn register(
        &mut self,
        base_dir: &str,
        cwd: &str,
        session_ids: Vec<String>,
    ) -> String {
        let hash = compute_cwd_hash(cwd);
        let composite_id = format!("{base_dir}::{hash}");

        let entry = self.entries.entry(composite_id.clone()).or_insert_with(|| {
            SubprojectEntry {
                cwd: cwd.to_string(),
                session_ids: HashSet::new(),
            }
        });

        for id in session_ids {
            entry.session_ids.insert(id);
        }

        composite_id
    }

    /// Get the session filter for a composite project ID.
    /// Returns None if the project ID is not composite.
    pub fn get_session_filter(&self, project_id: &str) -> Option<&HashSet<String>> {
        self.entries.get(project_id).map(|e| &e.session_ids)
    }

    /// Check if a project ID is a composite ID.
    pub fn is_composite(project_id: &str) -> bool {
        project_id.contains("::")
    }

    /// Clear the registry.
    pub fn clear(&mut self) {
        self.entries.clear();
    }
}

/// Compute the 8-character hex hash of a cwd string.
fn compute_cwd_hash(cwd: &str) -> String {
    let mut hasher = Sha256::new();
    hasher.update(cwd.as_bytes());
    let result = hasher.finalize();
    hex_encode(&result[..4]) // 4 bytes = 8 hex chars
}

fn hex_encode(bytes: &[u8]) -> String {
    bytes.iter().map(|b| format!("{b:02x}")).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_register_and_get_filter() {
        let mut registry = SubprojectRegistry::new();
        let id = registry.register(
            "-Users-name-project",
            "/Users/name/project",
            vec!["sess1".to_string(), "sess2".to_string()],
        );

        assert!(id.starts_with("-Users-name-project::"));
        assert_eq!(id.len(), "-Users-name-project::".len() + 8);

        let filter = registry.get_session_filter(&id).unwrap();
        assert!(filter.contains("sess1"));
        assert!(filter.contains("sess2"));
    }

    #[test]
    fn test_is_composite() {
        assert!(SubprojectRegistry::is_composite(
            "-Users-name-project::abcdef01"
        ));
        assert!(!SubprojectRegistry::is_composite("-Users-name-project"));
    }

    #[test]
    fn test_consistent_hash() {
        let hash1 = compute_cwd_hash("/Users/name/project");
        let hash2 = compute_cwd_hash("/Users/name/project");
        assert_eq!(hash1, hash2);
        assert_eq!(hash1.len(), 8);
    }

    #[test]
    fn test_different_cwds_different_hashes() {
        let hash1 = compute_cwd_hash("/Users/name/project1");
        let hash2 = compute_cwd_hash("/Users/name/project2");
        assert_ne!(hash1, hash2);
    }

    #[test]
    fn test_clear() {
        let mut registry = SubprojectRegistry::new();
        registry.register(
            "-Users-name-project",
            "/Users/name/project",
            vec!["sess1".to_string()],
        );
        registry.clear();
        assert!(registry.entries.is_empty());
    }
}

File: internal/discovery/subproject_registry.go
package discovery

// Subproject registry — tracks composite project IDs for multi-cwd projects.
// Mirrors src-tauri/src/discovery/subproject_registry.rs.
//
// When sessions in the same encoded directory have different `cwd` values they
// are split into separate projects with composite IDs: `{encodedPath}::{hash}`.
// The registry is a plain struct with a sync.Mutex — injected by the caller,
// never a package-level singleton.

import (
	"crypto/sha256"
	"fmt"
	"sync"
)

// SubprojectEntry holds the set of session IDs that belong to one subproject.
type SubprojectEntry struct {
	SessionIDs map[string]struct{}
}

// SubprojectRegistry maps composite project IDs to their session-ID sets.
// All exported methods are safe for concurrent use.
type SubprojectRegistry struct {
	mu      sync.Mutex
	entries map[string]*SubprojectEntry
}

// NewSubprojectRegistry returns an initialised, empty registry.
func NewSubprojectRegistry() *SubprojectRegistry {
	return &SubprojectRegistry{
		entries: make(map[string]*SubprojectEntry),
	}
}

// Register records the given sessionIDs under a composite ID derived from
// baseDir and cwd, returning the composite ID.
// Mirrors subproject_registry::SubprojectRegistry::register.
func (r *SubprojectRegistry) Register(baseDir, cwd string, sessionIDs []string) string {
	hash := computeCWDHash(cwd)
	compositeID := fmt.Sprintf("%s::%s", baseDir, hash)

	// nil-safe: a nil registry (the frontend passes null) still returns a valid
	// composite ID, it just doesn't persist the membership.
	if r == nil {
		return compositeID
	}

	r.mu.Lock()
	defer r.mu.Unlock()

	entry, ok := r.entries[compositeID]
	if !ok {
		entry = &SubprojectEntry{SessionIDs: make(map[string]struct{})}
		r.entries[compositeID] = entry
	}
	for _, id := range sessionIDs {
		entry.SessionIDs[id] = struct{}{}
	}
	return compositeID
}

// GetSessionFilter returns the set of session IDs for a composite project ID,
// or nil if the project ID is not composite / not registered.
// Mirrors subproject_registry::SubprojectRegistry::get_session_filter.
func (r *SubprojectRegistry) GetSessionFilter(projectID string) map[string]struct{} {
	if r == nil {
		return nil // no registry → no subproject filter
	}
	r.mu.Lock()
	defer r.mu.Unlock()

	entry, ok := r.entries[projectID]
	if !ok {
		return nil
	}
	return entry.SessionIDs
}

// Clear removes all entries. Mirrors subproject_registry::SubprojectRegistry::clear.
func (r *SubprojectRegistry) Clear() {
	if r == nil {
		return
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.entries = make(map[string]*SubprojectEntry)
}

// computeCWDHash returns the first 4 bytes of SHA-256(cwd) as 8 lowercase hex chars.
// Matches the Rust sha2::Sha256 output exactly.
func computeCWDHash(cwd string) string {
	sum := sha256.Sum256([]byte(cwd))
	return fmt.Sprintf("%02x%02x%02x%02x", sum[0], sum[1], sum[2], sum[3])
}
